src/batch.jl

```diff
@@ -0,0 +1,194 @@
+"""
+    BatchIntegrand(f!, y::AbstractVector, x::AbstractVector, max_batch::Integer)
+    BatchIntegrand(f!, y::AbstractVector, [x::AbstractVector]; max_batch=typemax(Int))
+    BatchIntegrand{Y,X}(f!; max_batch=typemax(Int)) where {Y,X}
+    BatchIntegrand{Y}(f!; max_batch=typemax(Int)) where {Y}
+
+Constructor for a `BatchIntegrand` accepting an integrand of the form `f!(y,x) = y .= f.(x)`
+that can evaluate the integrand at multiple quadrature nodes using, for example, threads,
+the GPU, or distributed memory. The arguments `y`, and optionally `x`, are pre-allocated
+buffers of the correct element type for the domain and range of `f!`. Additionally, they
+must be `resize!`-able since the number of evaluation points may vary between calls to `f!`.
+Alternatively, the element types of those buffers, `Y` and optionally `X`, may be passed as
+type parameters to the constructor. The `max_batch` keyword roughly limits the number of
+nodes passed to the integrand, though at least `4*order+2` nodes will be used by the GK
+rule.
+"""
```

Comment:
Maybe add a comment here that if `X` is not specified, it internally allocates a new `BatchIntegrand` inside `quadgk` based on the domain types, and that this internal `BatchIntegrand` re-uses the `y` buffer but allocates a new `x` buffer`.  So, if you want to re-use the `x` buffer between calls, you must supply `{Y,X}` or pass `y,x` explicitly.